core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+                // as that'll always be index 0:
+                Some(match self.primary_key {
+                    Some(..) => id + 1,
+                    None => id,
+                })
+            })
+    }
+
+    // Selects a prefix of `key` without the epoch or sequence number.
+    //
+    // The RocksDB docs state the following:
+    // > If non-nullptr, use the specified function to determine the
+    // > prefixes for keys.  These prefixes will be placed in the filter.
+    // > Depending on the workload, this can reduce the number of read-IOP
+    // > cost for scans when a prefix is passed via ReadOptions to
+    // > db.NewIterator(). For prefix filtering to work properly,
+    // > "prefix_extractor" and "comparator" must be such that the following
+    // > properties hold:
+    //
+    // > 1) key.starts_with(prefix(key))
+    // > 2) Compare(prefix(key), key) <= 0.
+    // > 3) If Compare(k1, k2) <= 0, then Compare(prefix(k1), prefix(k2)) <= 0
+    // > 4) prefix(prefix(key)) == prefix(key)
+    //
+    // NOTE(ekmartin): Encoding the key size in the key increases the total size with 4 bytes.
+    // If we really wanted to avoid this while still maintaining the same serialization scheme
+    // we could do so by figuring out how many bytes our bincode serialized Vec<DataType> takes
+    // up here in transform_fn. Example:
+    // [DataType::Int(1), DataType::BigInt(10)] would be serialized as:
+    // 2u64 (vec length), 0u32 (enum variant), 1i32 (value), 1u32 (enum variant), 1i64 (value)
+    // By stepping through the serialized bytes and checking each enum variant we would know
+    // when we reached the end, and could then with certainty say whether we'd already
+    // prefix transformed this key before or not
+    // (without including the byte size of Vec<DataType>).
+    fn transform_fn(key: &[u8]) -> Vec<u8> {
+        if key == META_KEY {
+            return Vec::from(key);
+        }
+
+        // IndexID is a u32, so bincode uses 4 bytes to serialize it (which we'll skip past):
+        let start = 4;
+        // We encoded the size of the key itself with a u64, which bincode uses 8 bytes to encode:
+        let size_offset = start + 8;
+        let key_size: u64 = bincode::deserialize(&key[start..size_offset]).unwrap();
+        let prefix_len = size_offset + key_size as usize;
+        let mut bytes = Vec::from(key);
+
+        // Strip away the IndexEpoch and IndexSeq if we haven't already done so:
+        if key.len() > prefix_len {
+            bytes.truncate(prefix_len);
+        }
+
+        bytes
+    }
+
+    // Decides which keys the prefix transform should apply to.
+    fn in_domain_fn(key: &[u8]) -> bool {
+        key != META_KEY
+    }
+
+    // A key is built up of five components:
+    //  * `index_id`
+    //     Uniquely identifies the index in self.indices.
+    //  * `row_size`
+    //      The byte size of `row` when serialized with binode.
```

Comment:
Maybe call these `key_` instead of `row_` given that they don't actually hold "rows".